src/interruptible/thread_data.h

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INTERRUPTIBLE_THREAD_DATA_H
+#define BITCOIN_INTERRUPTIBLE_THREAD_DATA_H
+
+#include <atomic>
+#include <assert.h>
+#include <mutex>
+
+namespace interruptible
+{
+namespace detail
+{
+class thread_data
+{
+public:
+    thread_data() noexcept = default;
+    thread_data(thread_data&& t) = default;
+    thread_data& operator=(thread_data&& t) = default;
+    ~thread_data() = default;
+
+    thread_data(const thread_data&) = delete;
+    thread_data& operator=(const thread_data&) = delete;
+
+    template <typename T, typename... Args>
+    void start_thread(T&& func, Args&&... args)
+    {
+        m_thread = std::thread(std::forward<T>(func), std::forward<Args>(args)...);
+    }
+
+    void set_interrupted()
+    {
+        m_interrupted.store(true, std::memory_order_relaxed);
```

Comment:
I've been thinking about this... I think it was actually fine as is. The store and the notify don't need to be ordered relative to each other because the release of the lock at the exit of the scope is synchronized-with the acquire of the lock that `wait` must perform before returning.